core/git-access.c

```diff
@@ -812,188 +813,69 @@ static struct git_repository *get_remote_repo(const char *localdir, const char *
 {
-	int flen, blen, ret;
-	int offset = 1;
 	struct stat st;
 	git_repository *repo;
-	char *loc, *branch;
-
-	flen = strlen(filename);
-	if (!flen || filename[--flen] != ']')
-		return NULL;
-
-	/*
-	 * Special-case "file://", and treat it as a local
-	 * repository since libgit2 is insanely slow for that.
-	 */
-	if (!strncmp(filename, "file://", 7)) {
-		filename += 7;
-		flen -= 7;
-	}
-
-	/* Find the matching '[' */
-	blen = 0;
-	while (flen && filename[--flen] != '[')
-		blen++;
+	int ret;
+	const char *localdir;
 
-	/* Ignore slashes at the end of the repo name */
-	while (flen && filename[flen-1] == '/') {
-		flen--;
-		offset++;
+	if (user && user[0]) {
+		if (prefs.cloud_storage_email_encoded)
+			free((void *)prefs.cloud_storage_email_encoded);
+		prefs.cloud_storage_email_encoded = strdup(user);
 	}
 
-	if (!flen)
-		return NULL;
-
-	/*
-	 * This is the "point of no return": the name matches
-	 * the git repository name rules, and we will no longer
-	 * return NULL.
-	 *
-	 * We will either return "dummy_git_repository" and the
-	 * branch pointer will have the _whole_ filename in it,
-	 * or we will return a real git repository with the
-	 * branch pointer being filled in with just the branch
-	 * name.
-	 *
-	 * The actual git reading/writing routines can use this
-	 * to generate proper error messages.
-	 */
-	*branchp = filename;
-	loc = format_string("%.*s", flen, filename);
-	if (!loc)
-		return dummy_git_repository;
-
-	branch = format_string("%.*s", blen, filename + flen + offset);
-	if (!branch) {
-		free(loc);
-		return dummy_git_repository;
-	}
-
-	if (dry_run) {
-		*branchp = branch;
-		*remote = loc;
-		return dummy_git_repository;
-	}
-	repo = is_remote_git_repository(loc, branch);
-	if (repo) {
-		if (remote)
-			*remote = loc;
-		else
-			free(loc);
-		*branchp = branch;
-		return repo;
+	/* remember if the current git storage we are working on is our cloud storage
+	 * this is used to create more user friendly error message and warnings */
+	is_subsurface_cloud = is_cloud;
+	is_local_cloud = is_cloud && !is_remote;
+
+	if (is_cloud || is_remote) {
+		/*
+		 * This turns a remote repository into a local one,
+		 * if possible.
+		 *
+		 * We now create the SHA1 hash of the whole thing,
+		 * including the branch name. That will be our unique
+		 * unique local repository name.
+		 *
+		 * NOTE! We will create a local repository per branch,
+		 * because
+		 *
+		 *  (a) libgit2 remote tracking branch support seems to
+		 *      be a bit lacking
+		 *  (b) we'll actually check the branch out so that we
+		 *      can do merges etc too.
+		 *
+		 * so even if you have a single remote git repo with
+		 * multiple branches for different people, the local
+		 * caches will sadly force that to split into multiple
+		 * individual repositories.
+		 */
+		localdir = get_local_dir(loc, branch);
+		if (!localdir)
+			return dummy_git_repository;
+
+		return get_remote_repo(localdir, loc, branch);
```

Comment:
Again, this looks like too much change in one commit. I don't understand why this code is moving and how it affects everything else.

Reply:
As above - the parsing needed to tell whether this is a remote git repository or whether there's a username in the URL is now done outside of the git machinery. Therefore, the whole parsing in "is_remote_git_repository()" is simply replaced by an input parameters and the rest of the function is placed into the calling function. Also, there is no need for a "dry run" option.